lib/Support/ValueMapper.cpp

```diff
@@ -0,0 +1,66 @@
+//===- ValueMapper.cpp - Support for mapping SSA values ---------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides support for mapping SSA values between two domains.
+// Provided a BackedgeBuilder, the ValueMapper supports mappings between
+// GraphRegions, creating Backedges in cases of 'get'ing mapped values which are
+// yet to be 'set'.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/ValueMapper.h"
+
+using namespace mlir;
+using namespace circt;
+mlir::Value ValueMapper::get(Value from, TypeTransformer typeTransformer) {
+  if (mapping.count(from) == 0) {
+    assert(bb && "Trying to 'get' a mapped value without any value set. No "
+                 "BackedgeBuilder was provided, so cannot provide any mapped "
+                 "SSA value!");
+    // Create a backedge which will be resolved at a later time once all
+    // operands are created.
+    mapping[from] = bb->get(typeTransformer(from.getType()));
+  }
+  auto operandMapping = mapping[from];
+  Value mappedOperand;
+  if (auto *v = std::get_if<Value>(&operandMapping))
```

Comment:
Can't set Backedges be used for this rather than a variant around a backedge and a value?  (It seems like Backedge already has the value or not yet property you want).